src/ui/features/referral-program/ReferralProgramService.background.ts

```diff
@@ -0,0 +1,132 @@
+import { emitter } from 'src/background/events';
+import { isReadonlyContainer } from 'src/shared/types/validators';
+import type { Wallet } from 'src/shared/types/Wallet';
+import type { WalletContainer } from 'src/shared/types/WalletContainer';
+import { INTERNAL_SYMBOL_CONTEXT } from 'src/background/Wallet/Wallet';
+import { invariant } from 'src/shared/invariant';
+import { ZerionAPI } from 'src/modules/zerion-api/zerion-api.background';
+import { normalizeAddress } from 'src/shared/normalizeAddress';
+import { readReferrer, saveReferrer } from './shared/storage';
+
+interface Options {
+  getWallet: () => Wallet;
+}
+
+class ReferralProgramService {
+  private options: Options | null = null;
+
+  initialize(options: Options) {
+    this.options = options;
+    emitter.on('walletCreated', this.applyReferralCodeToWallet.bind(this));
+  }
+
+  async signReferralMessage({
+    address,
+    referralCode,
+  }: {
+    address: string;
+    referralCode: string;
+  }) {
+    invariant(this.options, "Options aren't expected to become null");
+    const walletFacade = this.options.getWallet();
+
+    const signer = walletFacade.getOfflineSigner();
```

Comment:
Probably, we can do it with .personalSign method

Reply:
I realize I made a big mistake by signing a "referral wallet" message using the current wallet address as a signer